File: src/main/java/skytils/skytilsmod/core/UpdateChecker.java
/*
 * Skytils - Hypixel Skyblock Quality of Life Mod
 * Copyright (C) 2021 Skytils
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package skytils.skytilsmod.core;

import com.google.gson.JsonObject;
import kotlin.Unit;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.GuiMainMenu;
import net.minecraftforge.client.event.GuiOpenEvent;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.versioning.DefaultArtifactVersion;
import skytils.skytilsmod.Skytils;
import skytils.skytilsmod.gui.UpdateGui;
import skytils.skytilsmod.utils.APIUtil;

import java.io.*;

public class UpdateChecker {

    private final static Minecraft mc = Minecraft.getMinecraft();
    private final static UpdateGetter updateGetter = new UpdateGetter();

    static {
        Thread thread = new Thread(updateGetter);
        thread.start();
        try {
            thread.join();
        } catch (InterruptedException ex) {
            ex.printStackTrace();
        }
    }

    @SubscribeEvent
    public void onGuiOpen(GuiOpenEvent e) {
        if (!(e.gui instanceof GuiMainMenu)) return;
        if (updateGetter.updateObj == null) return;
        try {
/*            Notifications notifs = Notifications.INSTANCE;
            notifs.pushNotification("New Skytils Version Available", "Click here to download", () -> {
                Skytils.displayScreen = new UpdateGui();
                return Unit.INSTANCE;
            });*/
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    public static String getUpdateDownloadURL() {
        return updateGetter.updateObj.get("assets").getAsJsonArray().get(0).getAsJsonObject().get("browser_download_url").getAsString();
    }

    public static String getJarNameFromUrl(String url) {
        String[] sUrl = url.split("/");
        return sUrl[sUrl.length - 1];
    }

    public static void scheduleCopyUpdateAtShutdown(String jarName) {
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            try {
                System.out.println("Attempting to apply Skytils update.");
                File oldJar = Skytils.jarFile;

                if (oldJar == null || !oldJar.exists() || oldJar.isDirectory()) {
                    System.out.println("Old jar file not found.");
                    return;
                }

                File newJar = new File(new File(Skytils.modDir, "updates"), jarName);
                copyFile(newJar, oldJar);
                newJar.delete();
                System.out.println("Successfully applied Skytils update.");
            } catch (Throwable ex) {
                ex.printStackTrace();
            }
        }));
    }

    /**
     * Taken from Wynntils under GNU Affero General Public License v3.0
     * Modified to perform faster
     * https://github.com/Wynntils/Wynntils/blob/development/LICENSE
     * @author Wynntils
     * Copy a file from a location to another
     *
     * @param sourceFile The source file
     * @param destFile Where it will be
     */
    public static void copyFile(File sourceFile, File destFile) throws IOException {
        if (destFile == null || !destFile.exists()) {
            destFile = new File(new File(sourceFile.getParentFile(), "mods"), "Skytils.jar");
            sourceFile.renameTo(destFile);
            return;
        }

        InputStream source = null;
        OutputStream dest = null;
        try {
            source = new FileInputStream(sourceFile);
            dest = new FileOutputStream(destFile);
            byte[] buffer = new byte[1024];
            int length;
            while ((length = source.read(buffer)) > 0) {
                dest.write(buffer, 0, length);
            }
        } finally {
            source.close();
            dest.close();
        }
    }

    private static class UpdateGetter implements Runnable {
        private volatile JsonObject updateObj = null;

        /**
         * Modified version from Danker's Skyblock Mod, taken under GPL 3.0 license.
         * https://github.com/bowser0000/SkyblockMod/blob/master/LICENSE
         * @author bowser0000
         */
        @Override
        public void run() {
            System.out.println("Checking for updates...");

            JsonObject latestRelease = Skytils.config.updateChannel == 1 ? APIUtil.getArrayResponse("https://api.github.com/repos/Skytils/SkytilsMod/releases").get(0).getAsJsonObject() : APIUtil.getJSONResponse("https://api.github.com/repos/Skytils/SkytilsMod/releases/latest");
            String latestTag = latestRelease.get("tag_name").getAsString();
            DefaultArtifactVersion currentVersion = new DefaultArtifactVersion(Skytils.VERSION);
            DefaultArtifactVersion latestVersion = new DefaultArtifactVersion(latestTag.substring(1));

            if (latestTag.contains("pre") || Skytils.VERSION.contains("pre") && currentVersion.compareTo(latestVersion) >= 0) {
                double currentPre = 0;
                double latestPre = 0;
                if (Skytils.VERSION.contains("pre")) {
                    currentPre = Double.parseDouble(Skytils.VERSION.substring(Skytils.VERSION.indexOf("pre") + 3));
                }

                if (latestTag.contains("pre")) {
                    latestPre = Double.parseDouble(latestTag.substring(latestTag.indexOf("pre") + 3));
                }

                if (latestPre > currentPre || (latestPre == 0 && currentVersion.compareTo(latestVersion) == 0)) {
                    updateObj = latestRelease;
                }
            } else if (currentVersion.compareTo(latestVersion) < 0) {
                updateObj = latestRelease;
            }
        }

    }
}

File: src/main/kotlin/skytils/skytilsmod/core/UpdateChecker.kt
/*
 * Skytils - Hypixel Skyblock Quality of Life Mod
 * Copyright (C) 2022 Skytils
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package skytils.skytilsmod.core

import com.google.gson.JsonObject
import net.minecraft.client.gui.GuiMainMenu
import net.minecraft.util.Util
import net.minecraftforge.client.event.GuiOpenEvent
import net.minecraftforge.fml.common.eventhandler.EventPriority
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.versioning.DefaultArtifactVersion
import org.apache.hc.client5.http.classic.methods.HttpGet
import org.apache.hc.core5.http.io.entity.EntityUtils
import skytils.skytilsmod.Skytils
import skytils.skytilsmod.gui.RequestUpdateGui
import skytils.skytilsmod.gui.UpdateGui
import skytils.skytilsmod.utils.APIUtil
import skytils.skytilsmod.utils.Utils
import java.awt.Desktop
import java.io.File
import java.net.URL
import kotlin.concurrent.thread

object UpdateChecker {
    val updateGetter = UpdateGetter()
    val updateDownloadURL: String
        get() = updateGetter.updateObj!!["assets"].asJsonArray[0].asJsonObject["browser_download_url"].asString

    fun getJarNameFromUrl(url: String): String {
        return url.split(Regex("/")).last()
    }

    fun scheduleCopyUpdateAtShutdown(jarName: String) {
        Runtime.getRuntime().addShutdownHook(Thread {
            try {
                println("Attempting to apply Skytils update.")
                val oldJar = Skytils.jarFile
                if (oldJar == null || !oldJar.exists() || oldJar.isDirectory) {
                    println("Old jar file not found.")
                    return@Thread
                }
                println("Copying updated jar to mods.")
                val newJar = File(File(Skytils.modDir, "updates"), jarName)
                println("Copying to mod folder")
                val newLocation = File(oldJar.parent, "${if (oldJar.name.startsWith("!")) "!" else ""}${jarName}")
                newLocation.createNewFile()
                newJar.copyTo(newLocation, true)
                newJar.delete()
                if (oldJar.delete()) {
                    println("successfully deleted the files. skipping install tasks")
                    return@Thread
                }
                println("Running delete task")
                val taskFile = File(File(Skytils.modDir, "updates"), "tasks").listFiles()?.last()
                if (taskFile == null) {
                    println("Task doesn't exist")
                    return@Thread
                }
                val runtime = Utils.getJavaRuntime()
                if (Util.getOSType() == Util.EnumOS.OSX) {
                    val sipStatus = Runtime.getRuntime().exec("csrutil status")
                    sipStatus.waitFor()
                    if (!sipStatus.inputStream.use { it.bufferedReader().readText() }
                            .contains("System Integrity Protection status: disabled.")) {
                        println("SIP is NOT disabled, opening Finder.")
                        Desktop.getDesktop().open(oldJar.parentFile)
                        return@Thread
                    }
                }
                println("Using runtime $runtime")
                Runtime.getRuntime().exec("\"$runtime\" -jar \"${taskFile.absolutePath}\" \"${oldJar.absolutePath}\"")
                println("Successfully applied Skytils update.")
            } catch (ex: Throwable) {
                println("Failed to apply Skytils Update.")
                ex.printStackTrace()
            }
        })
    }

    fun downloadDeleteTask() {
        Thread {
            println("Checking for Skytils install task...")
            val taskDir = File(File(Skytils.modDir, "updates"), "tasks")
            // TODO Make this dynamic and fetch latest one or something
            val url =
                "https://github.com/Skytils/SkytilsMod-Data/releases/download/files/SkytilsInstaller-1.1.1.jar"
            val taskFile = File(taskDir, getJarNameFromUrl(url))
            if (taskDir.mkdirs() || taskFile.createNewFile()) {
                println("Downloading Skytils delete task.")
                val client = APIUtil.builder.build()

                val req = HttpGet(URL(url).toURI())
                val res = client.execute(req)
                if (res.code != 200) {
                    println("Downloading delete task failed!")
                } else {
                    println("Writing Skytils delete task.")
                    res.entity.writeTo(taskFile.outputStream())
                    EntityUtils.consume(res.entity)
                    client.close()
                    println("Delete task successfully downloaded!")
                }
            }
        }.start()
    }

    init {
        try {
            thread(block = updateGetter::run).join()
        } catch (ex: InterruptedException) {
            ex.printStackTrace()
        }
    }

    @SubscribeEvent(priority = EventPriority.HIGHEST)
    fun onGuiOpen(e: GuiOpenEvent) {
        if (e.gui !is GuiMainMenu) return
        if (updateGetter.updateObj == null) return
        if (UpdateGui.complete) return
        Skytils.displayScreen = RequestUpdateGui()
    }

    class UpdateGetter : Runnable {
        @Volatile
        var updateObj: JsonObject? = null

        override fun run() {
            println("Checking for updates...")
            val latestRelease = when (Skytils.config.updateChannel) {
                2 -> APIUtil.getJSONResponse(
                    "https://api.github.com/repos/Skytils/SkytilsMod/releases/latest"
                )
                1 -> APIUtil.getArrayResponse(
                    "https://api.github.com/repos/Skytils/SkytilsMod/releases"
                )[0].asJsonObject
                else -> return println("Channel set as none")
            }
            val latestTag = latestRelease["tag_name"].asString
            val currentTag = Skytils.VERSION.substringBefore("-dev")

            val currentVersion = SkytilsVersion(currentTag)
            val latestVersion = SkytilsVersion(latestTag.substringAfter("v"))
            if (currentVersion < latestVersion) {
                updateObj = latestRelease
            }
        }
    }

    class SkytilsVersion(val versionString: String) : Comparable<SkytilsVersion> {

        companion object {
            val regex by lazy { Regex("^(?<version>[\\d.]+)-?(?<type>\\D+)?(?<typever>\\d+\\.?\\d*)?\$") }
        }

        private val matched by lazy { regex.find(versionString) }
        val isSafe by lazy { matched != null }

        val version by lazy { matched!!.groups["version"]!!.value }
        val versionArtifact by lazy { DefaultArtifactVersion(matched!!.groups["version"]!!.value) }
        val specialVersionType by lazy {
            val typeString = matched!!.groups["type"]?.value ?: return@lazy UpdateType.RELEASE

            return@lazy UpdateType.values().find { typeString == it.prefix } ?: UpdateType.UNKNOWN
        }
        val specialVersion by lazy {
            if (specialVersionType == UpdateType.RELEASE) return@lazy null
            return@lazy matched!!.groups["typever"]?.value?.toDoubleOrNull()
        }

        override fun compareTo(other: SkytilsVersion): Int {
            if (!isSafe || !other.isSafe) return -1
            return if (versionArtifact.compareTo(other.versionArtifact) == 0) {
                if (specialVersionType.ordinal == other.specialVersionType.ordinal) {
                    (specialVersion ?: 0.0).compareTo(other.specialVersion ?: 0.0)
                } else other.specialVersionType.ordinal - specialVersionType.ordinal
            } else versionArtifact.compareTo(other.versionArtifact)
        }
    }

    enum class UpdateType(val prefix: String) {
        UNKNOWN("unknown"),
        RELEASE(""),
        RELEASECANDIDATE("RC"),
        PRERELEASE("pre"),
    }
}